blockmanager.go

```diff
@@ -534,7 +541,8 @@ func (b *blockManager) current() bool {
 func (b *blockManager) handleBlockMsg(bmsg *blockMsg) {
 	// If we didn't ask for this block then the peer is misbehaving.
 	blockSha := bmsg.block.Sha()
-	if _, ok := bmsg.peer.requestedBlocks[*blockSha]; !ok {
+
+	if p, exists := b.requestedBlocks[*blockSha]; !exists || p != bmsg.peer {
```

Comment:
This would have to be updated when changed per above.  Something like:
```Go
if p.ID() == bmsg.peer.ID() || b.hasRequestedBlock(p) {
// ...
}
```
Where that `b.hasRequestedBlock` is a new convenience func used throughout as needed and would look something like:
```Go
func (b *blockManager) hasRequestedBlock(p *peer.Peer) bool {
	requestedBlocks, ok := b.requestedTxns[p.ID()]
	if !ok {
		return false
	}

	_, ok := requestedBlocks[p.ID()]
	return ok
}
```